pkg/apis/resource/types.go

```diff
@@ -0,0 +1,466 @@
+	// delayed until a Pod gets scheduled that needs the Claim. The
+	// scheduler will consider all resource requirements of that Pod and
+	// trigger allocation for a node that fits the Pod.
+	AllocationModeWaitForFirstConsumer AllocationMode = "WaitForFirstConsumer"
+
+	// When a Claim has AllocationModeImmediate, allocation starts
+	// as soon as the Claim gets created. This is done without
+	// considering the needs of Pods that will use the Claim
+	// because those Pods are not known yet.
+	AllocationModeImmediate AllocationMode = "Immediate"
+)
+
+// ClaimStatus tracks whether the resource has been allocated and what
+// the resulting attributes are.
+type ClaimStatus struct {
+	// DriverName is a copy of the driver name from the Class at
+	// the time when allocation started. It's necessary to enable usage of
+	// the claim by kubelet in case that the Class got removed in
+	// the meantime. It also helps the resource driver to determine whether
+	// it needs to handle a claim that got marked for deletion.
+	DriverName string
+
+	// Allocation is set by the resource driver once a resource has been
+	// allocated successfully. Nil indicates that the resource is not
+	// allocated.
+	Allocation *AllocationResult
+
+	// ReservedFor indicates which entities are currently allowed to use
+	// the claim. Usually those are Pods, but other objects are also
+	// possible as long as they exist as resources in the apiserver. The
+	// Claim controller in the kube-controller-manager checks for
+	// stale entries that reference non-existent objects and removes them.
+	//
+	// A scheduler must add a Pod that it is scheduling. This must be done
+	// in an atomic Claim update because there might be multiple
+	// schedulers working on different Pods that compete for access to the
+	// same Claim, the Claim might have been marked for
+	// deletion, or even been deallocated already.
+	//
+	// kubelet will check this before allowing a Pod to run because a a
+	// user might have selected a node manually without reserving resources
+	// or a scheduler might have missed that step, for example because it
+	// doesn't support dynamic resource allocation or the feature was
+	// disabled.
+	//
+	// The maximum size is 32 (= [ClaimReservedForMaxSize]).
+	ReservedFor []ClaimUserReference
+
+	// DeallocationRequested gets set by the scheduler when it detects the
+	// situation where pod scheduling cannot proceed because some claim was
+	// allocated for a node that cannot provide some other required
+	// resource. It can only be set for claims that are not currently
+	// reserved.
+	//
+	// The driver then needs to deallocate this claim and reset the field
+	// together with clearing the Allocation field. The scheduler then will
+	// try again.
+	//
+	// While DeallocationRequested is set, no new users may be added to
+	// ReservedFor.
+	DeallocationRequested bool
+}
+
+// ReservedForMaxSize is the maximum number of entries in
+// [ClaimStatus.ReservedFor].
```

Comment:
This [syntax] is new - are we doing that now?  I don't think so.  Also docs should reference fields by their jsonName not their GoName

Reply:
I think this particular type definition will only appear in the go doc output, but to be safe I switched to json names and removed the special go doc formatting.

Reply:
It almost feels like we need two versions of these docs: once for end-users, using JSON, and once for Go developers, with all the technical details that they need to correctly work with these APIs and go doc formatting... but not now.

Reply:
We need an IDL, but that's a story for another day.